### ISSUE TITLE
Should decreaseIndentPattern ever be used when decreaseNextIndentPattern was just triggered?

### ISSUE THREAD

**Issue by kbrose** [atom/atom#15926]
I have two proposals I would like feedback on regarding indentation regexes.

1. Modify the behavior of `decreaseIndentPattern` so that it would only be tested if the previous line did _not_ trigger `decreaseNextIndentPattern`. I believe that would only require changing these lines:
https://github.com/atom/atom/blob/c9c495792188693ca234e4ac08efcb83b0036cf7/src/tokenized-buffer.js#L155-L157
2. Add a new indentation regex `decreaseIndentIfNotPreviousPattern` (or something more pithy) that would behave the same as `decreaseIndentPattern` but only be tested if the previous line did _not_ trigger `decreaseNextIndentPattern`.

From my perspective, I would prefer option 1, but I'm worried there may be some languages where it _does_ make sense for `decreaseIndentPattern` to be triggered even when the previous line triggered `decreaseNextIndentPattern`.

Context: indenting python code in Atom has always been less than optimal, and this change would allow proper dedentation after statements like `return x` without improperly dedenting in the case an `else:` statement directly follows, as in

```python
def f(x):
    if x:
        return x # should hit decreaseNextIndentPattern
    else: # should not hit decreaseIndentPattern since previous line already dedented
        return 'something'
```

Ref https://github.com/atom/language-python/pull/212

**Comment by 50Wliu** [comment]
I am :+1: for implementing option 1.  I too can't think of a situation where both decreaseNextIndent and decreaseIndent should be reached.

**Comment by kbrose** [comment]
@50Wliu Thoughts on this with respect to `language-go` relying on this behavior in cases such as this:

```go
    switch i {
    case 1:
        fmt.Println(
            "one") // relies on triggering decreaseNextIndentRegex
    case 2: // relies on again triggering decreaseIndentRegex
```

I would really like to get perfect auto-indentation for python, but not at the cost of other languages. Although, reading the original PR for this makes me think it's not a killer feature for the go users:

> For this I created PR atom/atom#6808 but in order to be able to write a proper test, I need at least one language package to include the new regexp so I can call it in the test.

https://github.com/atom/language-go/pull/44

Perhaps option 2 isn't the worst thing in the world...

**Comment by 50Wliu** [comment]
Personally I think language-go is using `decreaseNextIndentRegex` incorrectly.  Though they have an [even stricter take on formatting](https://golang.org/doc/effective_go.html#formatting) than Python's [PEP-8](https://www.python.org/dev/peps/pep-0008/), I think that attempting to format code using the current auto-indentation system is very much out of its scope (in fact, the current system can't really do anything that well).

However to preserve backwards-compatibility, option 2 might be the best option.  I'm also waiting on some feedback from other Atom team members regarding this.

**Comment by kbrose** [comment]
@svanharmelen Hope you don't mind the ping, it looks like you were the one who originally added this so I'm curious to hear your thoughts.

> in fact, the current system can't really do anything that well

Do you have any bigger-picture ideas? Or do you think it's best to punt to packages?

Something I've been tossing around in my head is allowing language packages to specify an `autoindent` function that takes the buffer position of the cursor (or something similar) and when called outputs a list of `[(buffer_row_1, indent_change_1), (buffer_row_2, indent_change_2)]` modifications to the indentations (or something similar). But to *really* support indentations in a lot of languages we'd need to allow changes to be fractions of the indent size, as in

```python
def f(x, y, z,
      a, b, c):
    pass
```
where the indent after `z` is 1.5 indents (assuming soft indent of 4 spaces).

(This function is essentially what the [python-indent](https://github.com/DSpeckhals/python-indent) package does.)

**Comment by svanharmelen** [comment]
@kbrose don't mind the ping, but it's been so long since I used Atom and/or worked on it's codebase that I honestly feel I'm no longer qualified to comment on this.

Back when I worked on this, the previous implementation seemed (to me at least) to have a few shortcomings/bugs which I tried to improve/fix. But only to find that the current `auto-indentation` code of Atom is a very delicate and pretty nasty beast to work with.

So I agree the current code is by no means perfect, yet I didn't gave this any additional thoughts as I stopped using Atom quite a while ago (not too long after the related PR was merged)...
